mypy/checkexpr.py

```diff
@@ -1143,17 +1144,23 @@ def check_overload_call(self,
                                                  callable_name=callable_name,
                                                  object_type=object_type)
                 if not unioned_errors.is_errors():
-                    # Success! Stop early.
-                    return unioned_result
+                    # Success! But we need to see maybe normal procedure gives a narrower type.
+                    union_success = True
 
         # Step 3: If the union math fails, or if there was no union in the argument types,
         #         we fall back to checking each branch one-by-one.
         inferred_result = self.infer_overload_return_type(plausible_targets, args, arg_types,
                                                           arg_kinds, arg_names, callable_name,
                                                           object_type, context, arg_messages)
         if inferred_result is not None:
-            # Success! Stop early.
-            return inferred_result
+            # Success! Stop early by returning the best among normal and unioned.
+            if not union_success:
+                return inferred_result
+            else:
+                assert unioned_result is not None
+                if is_subtype(inferred_result[0], unioned_result[0]):
+                    return inferred_result
+                return unioned_result
```

Comment:
Actually, just had another thought -- I think it would be clearer (and a little more efficient) if we added logic to return early if `union_success` is true somewhere around here. Currently, we return early only if the normal algorithm also succeeds.

(The logic currently does the right thing because step 5 happens to do what we want, but step 4 and down are supposed to be failure cases.)